onadata/apps/logger/migrations/0034_set_require_auth_at_project_level.py

```diff
@@ -0,0 +1,164 @@
+# Generated by Django 3.2.15 on 2023-11-02 15:30
+from itertools import islice
+from urllib.parse import urlparse
+
+from django.conf import settings
+from django.db import migrations
+from redis import Redis
+
+
+CHUNK_SIZE = 2000
+
+redis_client = Redis.from_url(settings.ENKETO_REDIS_MAIN_URL)
+
+
+def restore_open_rosa_server_in_redis(apps, schema_editor):
+
+    if settings.SKIP_HEAVY_MIGRATIONS:
+        return
+
+    print(
+        """
+        This migration might take a while. If it is too slow, you may want to
+        re-run migrations with SKIP_HEAVY_MIGRATIONS=True and apply this one
+        manually from the django shell.
+        """
+    )
+
+    XForm = apps.get_model('logger', 'XForm')  # noqa
+
+    parsed_url = urlparse(settings.KOBOCAT_URL)
+    server_url = f"{settings.KOBOCAT_URL.rstrip('/')}"
+
+    xforms_iter = (
+        XForm.objects.filter(require_auth=False)
+        .values('id_string', 'user__username')
+        .iterator(chunk_size=CHUNK_SIZE)
+    )
+
+    while True:
+        xforms = list(islice(xforms_iter, CHUNK_SIZE))
+        if not xforms:
+            break
+        keys = []
+        for xform in xforms:
+            username = xform['user__username']
+            id_string = xform['id_string']
+            keys.append(f"or:{parsed_url.netloc}/{username},{id_string}|{username}")
```

Comment:
It took me a while to understand again but yes it is. 

Because the Enketo ID is based on the key/value `or:<domain_name>/<username>,<id_string>`  . KPI does not change this key whatever the permission of anonymous user. It only changes the `openRosaServer` property of the `hset` saved under the key `id:<value of key/value described 🆙 ` 

To sum-up, all the key/value `or:*` are (and will) containing the username. 

The pipe trick is to let `lua` set it in the `hset`.